src/tokenizer/tokenizer.py

```diff
@@ -843,6 +845,41 @@ def Split_Sentence(t: Optional[Tok] = None) -> Tok:
         return t
 
 
+class TokenStream:
+    """
+    Wrapper for token iterator allowing lookahead.
+    """
+
+    def __init__(self, token_it, *, lookahead_size=2):
+        """Initialize from token iterator."""
+        self.__it = token_it
+        if lookahead_size <= 0:
+            lookahead_size = 1
+        self.__lookahead = deque(maxlen=lookahead_size)
+        self.__max_lookahead = lookahead_size
+
+    def __next__(self):
+        if self.__lookahead:
+            return self.__lookahead.popleft()
+        return next(self.__it)
+
+    def __iter__(self):
+        return self
+
+    def __getitem__(self, i):
+        if 0 <= i < self.__max_lookahead:
+            l = len(self.__lookahead)
+            try:
+                while l <= i:
+                    # Extend deque to lookahead
+                    self.__lookahead.append(next(self.__it))
+                    l += 1
+                return self.__lookahead[i]
+            except StopIteration:
```

Comment:
Be careful that the StopIteration is not lost (it probably isn't, as next() probably raises StopIteration every time it's called on an empty iterator, but it needs to be confirmed).

Reply:
Yes, `next` on an empty iterator raises StopIteration